sage_acsv/helpers.py

```diff
@@ -179,6 +179,166 @@ def NewtonRecur(H, N):
+                            ]
+                        ) * dgdv.column(i) * dgdv.column(j)
+                     for H in Hs
+                    ]
+                )
+            ) for i in range(d-s)
+        ] for j in range(d-s)
+    ]
+
+    def _delta(l, k):
+        return 1 if l == k else 0
+
+    Hess = matrix(
+        [
+            [
+                sum(
+                    [
+                        r[k] * (
+                            -vs[i] * vs[j] * d2gdv2[i][j][k-(d-s)]*vs[k] - _delta(i,j)*dgdv[k-(d-s),j]*vs[k]*vs[i] \
+                                + vs[i]*vs[j]*dgdv[k-(d-s),i]*dgdv[k-(d-s),j]
+                        )/vs[k]**2 for k in range(d-s, d)
+                    ]
+                ) for i in range(d-s)
+            ] for j in range(d-s)
+        ] 
+    )
+    if subs:
+        return Hess.subs(subs)
+
+    return Hess
+
+def IsContributing(vs, pt, r, factors, c):
+    r"""Determines if critical point `pt` is contributing; that is, `r` is in the interior
+    of the scaled log-normal cone of `factors` at `pt`
+    
+    INPUT:
+
+    * ``vs`` -- A list of variables
+    * ``pt`` -- A point
+    * ``r`` -- A direction vector
+    * ``factors`` -- A list of factors of `H` for which `pt` vanishes
+    * ``c`` -- The co-dimension of the intersection of factors
+
+    OUTPUT:
+
+    If vs is contributing
+
+    EXAMPLES::
+
+    """
+    critical_subs = {v:point for v, point in zip(vs, pt)}
+    # Compute irreducible components of H that contain the point
+    vanishing_factors = list(
+        f for f in factors
+        if f.subs(critical_subs) == 0
+    )
+    for f in vanishing_factors:
+        if all([f.derivative(v).subs(critical_subs)==0 for v in vs]):
+            raise ACSVException("Critical point {p} lies in non-smooth part of component {f}".format(p=pt, f=f))
+
+    vkjs = []
+    for f in vanishing_factors:
+        for v in vs:
+            if f.derivative(v).subs(critical_subs)!=0:
+                vkjs.append(v)
+                break
+        else:
+            # In theory this shouldn't ever happen, now that we check the condition before
+            raise ACSVException("All partials of component {comp} vanish at point {p}".format(comp=vanishing_factors, p=pt))
+
+    normals = matrix(
+        list(
+            [AA(
+                f.derivative(v).subs(critical_subs) * critical_subs[v] / (
+                    critical_subs[vkj] * f.derivative(vkj).subs(critical_subs)
+                )
+            ) for v in vs] for vkj, f in zip(vkjs, vanishing_factors)
+        )
+    )
+    
+    polytope = Polyhedron(rays=normals)
+    if r not in polytope:
+        return False
+    elif any([r in f for f in polytope.faces(c-1)]):
+        # If r is in the boundary of the log normal cone, point is non-generic
+        raise ACSVException(
+            "Non-generic critical point found - {w} is contained in {dim}-dimensional stratum".format(w = str(pt), dim = len(vs)-c)

```

Comment:
I'd write this as an f-string, matter of taste to some degree:

```suggestion
            f"Non-generic critical point found - {pt} is contained in {len(vs) - c}-dimensional stratum"
```